target/ppc/fpu_helper.c

```diff
@@ -3368,84 +3368,199 @@ static inline target_ulong compute_fp_flags(float_status *excp_ptr)
 
 EXCP_CHECK(float32)
 EXCP_CHECK(float64)
-
 #undef EXCP_CHECK
+
 #define GET_VsrBF(A,I) A->VsrBF(I)
 #define GET_VsrHF(A,I) A->VsrHF(I)
 #define GET_VsrSF(A,I) A->VsrSF(I)
 #define GET_VsrDF(A,I) A->VsrDF(I)
+/* Defines that will never be called as they're in !FLOAT code */
+#define GET_VsrSBF(A,I) 0
+#define GET_VsrSHF(A,I) 0
+#define GET_VsrSSF(A,I) 0
+#define GET_VsrSDF(A,I) 0
+
+/* DUPLICATED FUNCTION FROM int_helper*/
+static inline void set_vscr_sat(CPUPPCState *env)
+{
+    /* The choice of non-zero value is arbitrary.  */
+    env->vscr_sat.u32[0] = 1;
+}
+
+/* int32 operations for XVI*GER */
+static int32_t int32_t_muladd (int32_t a, int32_t b, int32_t c,
+                               bool sat, void* env)
+{
+    int64_t result = ((int64_t)a * (int64_t)b);
+    result += (int64_t)c;
+    if (env && result > ((1LL << 31)-1)) {
+        set_vscr_sat(env);
+        return INT_MAX;
+    } else if (env && result < -(1LL << 31)) {
+        set_vscr_sat(env);
+        return INT_MIN;
+    } else {
+        return (int32_t) result;
+    }
+}
+
+static int32_t int32_t_add (int32_t a, int32_t b, void* env)
+{
+    int64_t result = ((int64_t)a);
+    result += (int64_t)b;
+    if (env && result > ((1LL << 31)-1)) {
+        set_vscr_sat(env);
+        return INT_MAX;
+    } else if (env && result < -(1LL << 31)) {
+        set_vscr_sat(env);
+        return INT_MIN;
+    } else {
+        return (int32_t)result;
+    }
+}
+
+/* CODE IN FLOAT SESSION */
+
+static int32_t int32_t_sub (int32_t a, int32_t b, void* env)
+{
+    g_assert_not_reached();
+    return 0;
+}
+
+static int32_t int32_t_mul (int32_t a, int32_t b, void* env)
+{
+    g_assert_not_reached();
+    return 0;
+}
+
+static inline target_ulong int32_t_check_muladd(int32_t a, int32_t b,
+                                                int32_t c, CPUPPCState *env)
+{
+    g_assert_not_reached();
+    return 0;
+}
+static inline target_ulong int32_t_check_add(int32_t a, int32_t b,
+                                             CPUPPCState *env)
+{
+    g_assert_not_reached();
+    return 0;
+}
+static inline target_ulong int32_t_compute_snan(int32_t a, CPUPPCState *env)
+{
+    g_assert_not_reached();
+    return 0;
+}
+
+#define GET_VsrN(A, I) ((a->VsrB((I)/2) >> ((((I)+1) % 2)*4)) & 0xF)
+#define GET_VsrB(A, I) A->VsrB(I)
+#define GET_VsrH(A, I) A->VsrH(I)
+#define GET_VsrSN(A, I) (GET_VsrN(A,I) | (((GET_VsrN(A,I) >> 3) & 1) ? (-1)^0xF : 0))
```

Comment:
Similarly,

```
#define GET_VsrSN(A, I) (int8_t)sextract32(A->VsrB((I) / 2), ((I) + 1) % 2, 4)
```

(note that we don't currently have ``sextract8``. Another option would be to add it.)